enn_zoo/enn_zoo/microrts/__init__.py

```diff
@@ -0,0 +1,360 @@
+            action["base_action"], CategoricalAction
+        ):
+            base_action_actors = action["base_action"].actors
+            base_actions = action["base_action"].actions
+        if "barrack_action" in action and isinstance(
+            action["barrack_action"], CategoricalAction
+        ):
+            barrack_action_actors = action["barrack_action"].actors
+            barrack_actions = action["barrack_action"].actions
+
+        response = self.client.gameStep(
+            unit_action_actors,
+            unit_actions,
+            base_action_actors,
+            base_actions,
+            barrack_action_actors,
+            barrack_actions,
+            0,
+        )
+
+        # self.client.render(False)
+        unit_action_actor_ids = np.array(response.observation[8])
+        unit_action_actor_masks = None
+        if len(unit_action_actor_ids) > 0:
+            unit_action_actor_masks = np.array(response.observation[9], dtype=np.bool8)
+        base_action_actor_ids = np.array(response.observation[10])
+        base_action_actor_masks = None
+        if len(base_action_actor_ids) > 0:
+            base_action_actor_masks = np.array(response.observation[11], dtype=np.bool8)
+        barrack_action_actor_ids = np.array(response.observation[12])
+        barrack_action_actor_masks = None
+        if len(barrack_action_actor_ids) > 0:
+            barrack_action_actor_masks = np.array(
+                response.observation[13], dtype=np.bool8
+            )
+        self.returns += response.reward
+
+        return Observation.from_entity_obs(
+            entities=self.generate_entities(response),
+            actions={
+                "unit_action": CategoricalActionMask(
+                    actor_ids=unit_action_actor_ids,  # type: ignore
+                    mask=unit_action_actor_masks,
+                ),
+                "base_action": CategoricalActionMask(
+                    actor_ids=base_action_actor_ids,  # type: ignore
+                    mask=base_action_actor_masks,
+                ),
+                "barrack_action": CategoricalActionMask(
+                    actor_ids=barrack_action_actor_ids,  # type: ignore
+                    mask=barrack_action_actor_masks,
+                ),
+            },
+            reward=response.reward @ self.reward_weight,
+            done=response.done[0],
+            end_of_episode_info=EpisodeStats(
+                length=self.step,
+                total_reward=float(self.reward_weight @ self.returns),
+                metrics=dict(
+                    zip(
+                        [f"charts/episodic_return/{item}" for item in self.rfs_names],
+                        self.returns,
+                    )
+                ),
+            )
+            if response.done[0]
+            else None,
+        )
+
+    def generate_entities(self, response: Any) -> Mapping[str, Optional[EntityObs]]:
+        entities: MutableMapping[str, Optional[EntityObs]] = {}
```

Comment:
could make this a bit more compact with something like:
```
for entity_type, observation in zip(["Resource", "Base", ...], response.observation):
  ...
```

Reply:
This is a great suggestion! Thank you!